File: bindings/python/examples/cancellation.py
"""Example: cancelling an in-progress stream from another thread.

cancel() is thread-safe. A watchdog thread can cancel the stream based
on a timeout; the main thread sees the cancellation on its next iteration.

Run: LD_LIBRARY_PATH=../../build python3 examples/cancellation.py
"""

import http.server
import json
import sys
import threading
import time
import urllib.parse

sys.path.insert(0, str(__import__("pathlib").Path(__file__).parent.parent))
from apiexec import Stream, ApiExecError


class MockHandler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        time.sleep(0.01)

        parsed = urllib.parse.urlparse(self.path)
        params = urllib.parse.parse_qs(parsed.query)
        page = int(params.get("cursor", ["0"])[0])

        # Infinite stream
        body = {"data": [{"id": page}], "next": str(page + 1)}
        response = json.dumps(body).encode()
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(response)))
        self.end_headers()
        self.wfile.write(response)

    def log_message(self, *args):
        pass


server = http.server.HTTPServer(("127.0.0.1", 0), MockHandler)
port = server.server_address[1]
threading.Thread(target=server.serve_forever, daemon=True).start()

config = json.dumps({"base_url": f"http://127.0.0.1:{port}"})
stream = Stream("generic_rest", config, '{"prefetch_depth": 0}')


def watchdog():
    time.sleep(0.2)
    print("\n[watchdog] cancelling stream")
    stream.cancel()


threading.Thread(target=watchdog, daemon=True).start()

count = 0
try:
    for _ in stream:
        count += 1
except ApiExecError as e:
    if e.code == -6:  # CANCELLED
        print(f"Stream cancelled cleanly after {count} batches")
    else:
        print(f"Error: {e}")

stream.close()
server.shutdown()

File: bindings/rust/examples/cancellation.rs
//! Example: cancelling an in-progress stream from another thread.
//!
//! Run: cargo run --example cancellation

use apiexec::{Error, Stream};
use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream};
use std::sync::atomic::{AtomicI32, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

fn handle_client(mut stream: TcpStream, counter: Arc<AtomicI32>) {
    let mut buf = [0u8; 4096];
    let _ = stream.read(&mut buf);
    thread::sleep(Duration::from_millis(10));

    let page = counter.fetch_add(1, Ordering::SeqCst);
    let body = format!(
        r#"{{"data":[{{"id":{}}}],"next":"{}"}}"#,  // infinite stream
        page,
        page + 1
    );
    let resp = format!(
        "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}",
        body.len(), body
    );
    let _ = stream.write_all(resp.as_bytes());
}

fn main() {
    let listener = TcpListener::bind("127.0.0.1:0").unwrap();
    let port = listener.local_addr().unwrap().port();
    let counter = Arc::new(AtomicI32::new(0));

    let counter_clone = counter.clone();
    thread::spawn(move || {
        for stream in listener.incoming() {
            if let Ok(s) = stream {
                handle_client(s, counter_clone.clone());
            }
        }
    });

    let config = format!(r#"{{"base_url":"http://127.0.0.1:{}"}}"#, port);
    let stream = Arc::new(Stream::new("generic_rest", &config, r#"{"prefetch_depth": 0}"#).unwrap());

    // Watchdog thread cancels after 200ms
    let stream_clone = stream.clone();
    thread::spawn(move || {
        thread::sleep(Duration::from_millis(200));
        println!("\n[watchdog] cancelling stream");
        stream_clone.cancel();
    });

    let mut count = 0;
    while stream.has_next() {
        match stream.next_batch(0) {
            Ok(_) => count += 1,
            Err(Error::Cancelled) => {
                println!("Stream cancelled after {} batches", count);
                return;
            }
            Err(e) => {
                println!("Error: {}", e);
                break;
            }
        }
    }
    println!("Stream ended after {} batches", count);
}
